File: js/src/codec.js
/**
 * v18/v20 binary codec reader — decode-only.
 *
 * Reads .v18 residual files: Header → LeafIndex → LeafBlobs → RecordStreams.
 * Port of proofcodec_verify/codec/v18_codec.py.
 *
 * Wire format (authoritative — from v18_codec.py, not CODEC_SPEC.md):
 *   Header: 112 bytes
 *   LeafIndexEntry: 20 bytes each
 *   LeafBlobHeader: 29 bytes (v18.0) or 41 bytes (v18.1)
 */

import { decodeBlock, decodeEmptyRun, RecordType } from './residual.js';
import { decodeUvarint } from './leb128.js';

export const MAGIC = new Uint8Array([0x43, 0x45, 0x47, 0x52]); // "CEGR"
export const DEFAULT_BLOCK_SIZE = 4096;

export const LabelCodec = { FIXED_2BIT: 0, HUFFMAN: 1, CONDITIONAL_1BIT: 2 };
export const PartitionMode = { LEAF_BLOCK: 0, LEAF_BLOCK_PRED: 1 };

/**
 * Parse the 112-byte V18 header.
 * @param {DataView} dv
 * @returns {Object}
 */
export function parseHeader(dv) {
  const magic = new Uint8Array(dv.buffer, dv.byteOffset, 4);
  if (magic[0] !== 0x43 || magic[1] !== 0x45 || magic[2] !== 0x47 || magic[3] !== 0x52) {
    throw new Error(`Invalid magic: ${String.fromCharCode(...magic)}`);
  }
  const versionMajor = dv.getUint16(4, true);
  return {
    versionMajor,
    versionMinor: dv.getUint16(6, true),
    flags: dv.getUint32(8, true),
    blockSize: dv.getUint32(12, true),
    numLeaves: dv.getUint32(16, true),
    partitionMode: dv.getUint8(20),
    labelCodec: dv.getUint8(21),
    nClasses: versionMajor >= 30 ? dv.getUint8(22) : 0,
    modelHash: new Uint8Array(dv.buffer, dv.byteOffset + 24, 32),
    syzygyHash: new Uint8Array(dv.buffer, dv.byteOffset + 56, 32),
    leafIndexOffset: Number(dv.getBigUint64(88, true)),
    totalPositions: Number(dv.getBigUint64(96, true)),
    totalMismatches: Number(dv.getBigUint64(104, true)),
  };
}

/**
 * Parse leaf index table.
 * @param {DataView} dv
 * @param {number} numLeaves
 * @param {number} offset
 * @returns {Array<{leafId: number, blobOffset: number, blobLen: number}>}
 */
export function parseLeafIndex(dv, numLeaves, offset) {
  const entries = [];
  for (let i = 0; i < numLeaves; i++) {
    const base = offset + i * 20;
    entries.push({
      leafId: dv.getUint32(base, true),
      blobOffset: Number(dv.getBigUint64(base + 4, true)),
      blobLen: Number(dv.getBigUint64(base + 12, true)),
    });
  }
  return entries;
}

/**
 * Parse a leaf blob header (29 bytes for v18.0).
 * @param {DataView} dv
 * @param {number} offset
 * @returns {[Object, number]} [header, newOffset]
 */
export function parseLeafBlobHeader(dv, offset) {
  const leafId = dv.getUint32(offset, true);
  const predCode = dv.getUint8(offset + 4);
  const nLeaf = dv.getUint32(offset + 5, true);
  const numBlocks = dv.getUint32(offset + 9, true);
  const kLeaf = dv.getUint32(offset + 13, true);
  const recordStreamLen = Number(dv.getBigUint64(offset + 17, true));
  const recordStreamCrc32 = dv.getUint32(offset + 25, true);
  return [{
    leafId,
    basePredLabel: predCode - 1,
    nLeaf,
    numBlocks,
    kLeaf,
    recordStreamLen,
    recordStreamCrc32,
  }, offset + 29];
}

/**
 * Decode a record stream into block data.
 * @param {Uint8Array} data
 * @param {number} numBlocks
 * @param {number} nLeaf
 * @param {number} blockSize
 * @param {number|null} leafPrediction
 * @param {number} nClasses - number of classes (0 = legacy 3-class WDL)
 * @returns {Array<[number, Object|null]>} [(blockId, blockData|null), ...]
 */
export function decodeRecordStream(data, numBlocks, nLeaf, blockSize, leafPrediction = null, nClasses = 0) {
  let offset = 0;
  let blockId = 0;
  const results = [];

  while (blockId < numBlocks && offset < data.length) {
    const recType = data[offset] & 0x3;

    if (recType === RecordType.EMPTY_RUN) {
      let runLen;
      [runLen, offset] = decodeEmptyRun(data, offset);
      for (let i = 0; i < runLen; i++) {
        results.push([blockId++, null]);
      }
    } else if (recType === RecordType.NON_EMPTY_BLOCK) {
      const nB = blockId < numBlocks - 1 ? blockSize : nLeaf - blockId * blockSize;
      let blockData, encoding;
      [blockData, encoding, offset] = decodeBlock(data, nB, offset, leafPrediction, nClasses);
      results.push([blockId++, blockData]);
    } else {
      throw new Error(`Unknown record type: ${recType}`);
    }
  }
  return results;
}

/**
 * Parse an entire v18 residual file.
 * @param {ArrayBuffer} buffer
 * @returns {Object} { header, leafIndex, leafBlobs }
 */
export function parseResidualFile(buffer) {
  const data = new Uint8Array(buffer);
  const dv = new DataView(buffer);

  const header = parseHeader(dv);
  const leafIndex = parseLeafIndex(dv, header.numLeaves, header.leafIndexOffset);

  const leafBlobs = new Map();
  for (const entry of leafIndex) {
    const [blobHeader, dataStart] = parseLeafBlobHeader(dv, entry.blobOffset);
    const recordStream = data.slice(dataStart, dataStart + blobHeader.recordStreamLen);
    leafBlobs.set(entry.leafId, { header: blobHeader, recordStream });
  }

  return { header, leafIndex, leafBlobs };
}

/**
 * Build a lookup cache from a parsed residual file.
 * @param {Object} file — from parseResidualFile
 * @returns {Map<number, Map<number, Map<number, number>>>} leafId → blockId → idxInBlock → label
 */
export function buildLookupCache(file) {
  const cache = new Map();

  for (const [leafId, blob] of file.leafBlobs) {
    const leafPred = file.header.labelCodec === LabelCodec.CONDITIONAL_1BIT
      ? blob.header.basePredLabel : null;

    const blocks = decodeRecordStream(
      blob.recordStream,
      blob.header.numBlocks,
      blob.header.nLeaf,
      file.header.blockSize,
      leafPred,
      file.header.nClasses,
    );

    const leafCache = new Map();
    for (const [bid, blockData] of blocks) {
      if (blockData && blockData.indices.length > 0) {
        const blockCache = new Map();
        for (let i = 0; i < blockData.indices.length; i++) {
          blockCache.set(blockData.indices[i], blockData.labels[i]);
        }
        leafCache.set(bid, blockCache);
      }
    }
    cache.set(leafId, leafCache);
  }
  return cache;
}

File: wasm/src/codec.rs
/// v18/v20 binary codec reader.
///
/// Wire format (from v18_codec.py — authoritative):
///   Header: 112 bytes
///   LeafIndexEntry: 20 bytes each
///   LeafBlobHeader: 29 bytes (v18.0)
use std::collections::HashMap;

use crate::residual::{decode_block, decode_empty_run, BlockData, RECORD_EMPTY_RUN, RECORD_NON_EMPTY_BLOCK};

pub const MAGIC: &[u8; 4] = b"CEGR";
pub const HEADER_SIZE: usize = 112;
pub const LEAF_INDEX_ENTRY_SIZE: usize = 20;
pub const LEAF_BLOB_HEADER_SIZE: usize = 29;

pub const LABEL_CODEC_FIXED_2BIT: u8 = 0;
pub const LABEL_CODEC_CONDITIONAL_1BIT: u8 = 2;

#[derive(Debug)]
pub struct V18Header {
    pub version_major: u16,
    pub version_minor: u16,
    pub block_size: u32,
    pub num_leaves: u32,
    pub partition_mode: u8,
    pub label_codec: u8,
    pub n_classes: u8,
    pub leaf_index_offset: u64,
    pub total_positions: u64,
    pub total_mismatches: u64,
}

#[derive(Debug)]
pub struct LeafBlobHeader {
    pub leaf_id: u32,
    pub base_pred_label: i8,
    pub n_leaf: u32,
    pub num_blocks: u32,
    pub k_leaf: u32,
    pub record_stream_len: u64,
    pub record_stream_crc32: u32,
}

/// Lookup cache: leaf_id → block_id → idx_in_block → label
pub type LookupCache = HashMap<u32, HashMap<u32, HashMap<u32, i8>>>;

fn read_u16_le(data: &[u8], off: usize) -> u16 {
    u16::from_le_bytes([data[off], data[off + 1]])
}

fn read_u32_le(data: &[u8], off: usize) -> u32 {
    u32::from_le_bytes([data[off], data[off + 1], data[off + 2], data[off + 3]])
}

fn read_u64_le(data: &[u8], off: usize) -> u64 {
    u64::from_le_bytes([
        data[off], data[off + 1], data[off + 2], data[off + 3],
        data[off + 4], data[off + 5], data[off + 6], data[off + 7],
    ])
}

/// Parse the 112-byte header.
pub fn parse_header(data: &[u8]) -> Result<V18Header, String> {
    if data.len() < HEADER_SIZE {
        return Err(format!("Header too short: {} < {}", data.len(), HEADER_SIZE));
    }
    if &data[0..4] != MAGIC {
        return Err(format!("Invalid magic: {:?}", &data[0..4]));
    }
    let version_major = read_u16_le(data, 4);
    Ok(V18Header {
        version_major,
        version_minor: read_u16_le(data, 6),
        block_size: read_u32_le(data, 12),
        num_leaves: read_u32_le(data, 16),
        partition_mode: data[20],
        label_codec: data[21],
        n_classes: if version_major >= 30 { data[22] } else { 0 },
        leaf_index_offset: read_u64_le(data, 88),
        total_positions: read_u64_le(data, 96),
        total_mismatches: read_u64_le(data, 104),
    })
}

/// Parse leaf index table entries.
pub fn parse_leaf_index(data: &[u8], num_leaves: u32, offset: usize) -> Vec<(u32, u64, u64)> {
    let mut entries = Vec::with_capacity(num_leaves as usize);
    for i in 0..num_leaves as usize {
        let base = offset + i * LEAF_INDEX_ENTRY_SIZE;
        let leaf_id = read_u32_le(data, base);
        let blob_offset = read_u64_le(data, base + 4);
        let blob_len = read_u64_le(data, base + 12);
        entries.push((leaf_id, blob_offset, blob_len));
    }
    entries
}

/// Parse a leaf blob header (29 bytes).
pub fn parse_leaf_blob_header(data: &[u8], offset: usize) -> (LeafBlobHeader, usize) {
    let leaf_id = read_u32_le(data, offset);
    let pred_code = data[offset + 4];
    let n_leaf = read_u32_le(data, offset + 5);
    let num_blocks = read_u32_le(data, offset + 9);
    let k_leaf = read_u32_le(data, offset + 13);
    let record_stream_len = read_u64_le(data, offset + 17);
    let record_stream_crc32 = read_u32_le(data, offset + 25);

    (
        LeafBlobHeader {
            leaf_id,
            base_pred_label: pred_code as i8 - 1,
            n_leaf,
            num_blocks,
            k_leaf,
            record_stream_len,
            record_stream_crc32,
        },
        offset + LEAF_BLOB_HEADER_SIZE,
    )
}

/// Decode a record stream into block data.
pub fn decode_record_stream(
    data: &[u8],
    num_blocks: u32,
    n_leaf: u32,
    block_size: u32,
    leaf_prediction: Option<i8>,
    n_classes: u8,
) -> Vec<(u32, Option<BlockData>)> {
    let mut offset = 0usize;
    let mut block_id = 0u32;
    let mut results = Vec::new();

    while block_id < num_blocks && offset < data.len() {
        let rec_type = data[offset] & 0x3;

        if rec_type == RECORD_EMPTY_RUN {
            let (run_len, new_off) = decode_empty_run(data, offset);
            offset = new_off;
            for _ in 0..run_len {
                results.push((block_id, None));
                block_id += 1;
            }
        } else if rec_type == RECORD_NON_EMPTY_BLOCK {
            let n_b = if block_id < num_blocks - 1 {
                block_size
            } else {
                n_leaf - block_id * block_size
            };
            let (block_data, _, new_off) = decode_block(data, n_b, offset, leaf_prediction, n_classes);
            offset = new_off;
            results.push((block_id, Some(block_data)));
            block_id += 1;
        } else {
            panic!("Unknown record type: {}", rec_type);
        }
    }
    results
}

/// Parse an entire v18 residual file and build a lookup cache.
pub fn parse_and_cache(data: &[u8]) -> Result<(V18Header, LookupCache), String> {
    let header = parse_header(data)?;
    let leaf_index = parse_leaf_index(data, header.num_leaves, header.leaf_index_offset as usize);

    let mut cache: LookupCache = HashMap::new();

    for (leaf_id, blob_offset, _blob_len) in &leaf_index {
        let (blob_header, data_start) = parse_leaf_blob_header(data, *blob_offset as usize);
        let rs_end = data_start + blob_header.record_stream_len as usize;
        let record_stream = &data[data_start..rs_end];

        let leaf_pred = if header.label_codec == LABEL_CODEC_CONDITIONAL_1BIT {
            Some(blob_header.base_pred_label)
        } else {
            None
        };

        let blocks = decode_record_stream(
            record_stream,
            blob_header.num_blocks,
            blob_header.n_leaf,
            header.block_size,
            leaf_pred,
            header.n_classes,
        );

        let mut leaf_cache: HashMap<u32, HashMap<u32, i8>> = HashMap::new();
        for (bid, block_data) in blocks {
            if let Some(bd) = block_data {
                if !bd.indices.is_empty() {
                    let mut block_cache = HashMap::new();
                    for (i, &idx) in bd.indices.iter().enumerate() {
                        block_cache.insert(idx, bd.labels[i]);
                    }
                    leaf_cache.insert(bid, block_cache);
                }
            }
        }
        cache.insert(*leaf_id, leaf_cache);
    }

    Ok((header, cache))
}
